### PR TITLE
Add support for fetching artifact attestations at the org, repo and user level

### PR DESCRIPTION
**Pull request by gillisandrew**
[GitHub artifact attestations is now generally available](https://github.blog/changelog/2024-06-25-artifact-attestations-is-generally-available/)

This pull request adds support for listing attestations hosted on GitHub for a given subject digest. Those attestations can then be verified using another library such as [sigstore-go](https://github.com/sigstore/sigstore-go). I've included a basic demo as an example.

### CODE REVIEW

File: example/go.mod
Comment (gmlewis) [inline example/go.mod:12]: Why is `github.com/sigstore/sigstore-go` pulling in an enormous boat load of dependencies below with it?

File: example/verifyartifact/main.go
Comment (gmlewis) [inline example/verifyartifact/main.go:33]: How does a user know where to get this?

File: example/verifyartifact/main.go
Comment (gmlewis) [inline example/verifyartifact/main.go:34]: How is this file generated?

File: github/attestations.go
Comment (gmlewis) [inline github/attestations.go:12]: All exported structs need Godoc-styled comments for the auto-generated docs with complete sentences and punctuation.

File: github/attestations.go
Comment (gmlewis) [inline github/attestations.go:17]:
```suggestion
	RepositoryID *int64             `json:"repository_id"`
```

File: github/attestations.go
Comment (gmlewis) [inline github/attestations.go:20]: Is there any reason why we wouldn't want to un-export this struct and return `[]*Attestation` as the first value from the newly-added methods?

I'm fine either way, it just seems like returning a slice might be more useful to end-users of this client library.

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @gillisandrew!
A few questions and tweaks, please.

File: example/go.mod
Comment (gillisandrew) [reply example/go.mod:12]: The library does a lot, including signing, hashing, provisioning of short-lived certs etc. If you'd prefer not to pull in so many indirect deps in the example module I can break it out into its own go module or host the example code in a different repo.

File: example/verifyartifact/main.go
Comment (gillisandrew) [reply example/verifyartifact/main.go:33]: It's written with the assumption that you've had some introduction to [how sigstore works](https://www.sigstore.dev/how-it-works).  GitHub presents these values online ([the code references this attestation](https://github.com/cli/cli/attestations/2543768)).

File: example/verifyartifact/main.go
Comment (gillisandrew) [reply example/verifyartifact/main.go:34]: The json file is a snapshot of the public good trusted root hosted by sigstore that underlies the fulcio CA that GitHub uses to sign the attestations in public repos. 

https://github.com/sigstore/root-signing
https://github.com/sigstore/fulcio

File: github/attestations.go
Comment (gillisandrew) [reply github/attestations.go:12]: Ok, I'll add some.

File: example/verifyartifact/main.go
Comment (gmlewis) [reply example/verifyartifact/main.go:33]: Perfect answer... thank you!
Can you please embed these answers (as `// comments`) in this file?

File: example/verifyartifact/main.go
Comment (gmlewis) [reply example/verifyartifact/main.go:34]: Thank you!  This would also be an extremely valuable comment to add to this file.

File: github/attestations.go
Comment (gillisandrew) [reply github/attestations.go:20]: I had considered it, however its a brand new endpoint and GitHub saw fit to return it as a JSON object instead of an array so it felt like unwrapping it might result in unnecessary breaking changes if GitHub starts including more data in the response.

Comment (gmlewis) [reply github/attestations.go:20]: OK, that makes total sense. Let's keep it then.

File: github/attestations.go
Comment (gmlewis) [inline github/attestations.go:21]: `json.RawMessage` is already a reference type, so the pointer is not necessary here.
Come to think of it, the `*int64` isn't necessary either. Sorry I didn't realize this before.

```suggestion
	Bundle       json.RawMessage `json:"bundle"`
	RepositoryID int64           `json:"repository_id"`
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @gillisandrew .
Please add unit tests for the newly-added methods.

File: github/orgs_attestations_test.go
Comment (gmlewis) [inline github/orgs_attestations_test.go:58]: Please do similarly to this in all new unit tests in order to increase code coverage.

```suggestion
	const methodName = "ListAttestations"

	testBadOptions(t, methodName, func() (err error) {
		_, _, err = client.Organizations.ListAttestations(ctx, "\n", "\n", &ListOptions{})
		return err
	})

	testNewRequestAndDoFailure(t, methodName, client, func() (*Response, error) {
		got, resp, err := client.Organizations.ListAttestations(ctx, "o", "digest", nil)
		if got != nil {
			t.Errorf("testNewRequestAndDoFailure %v = %#v, want nil", methodName, got)
		}
		return resp, err
	})
}
```

Comment (gmlewis) [reply github/orgs_attestations_test.go:72]: Please do similarly to this in all new unit tests in order to increase code coverage.

**Review by gmlewis — APPROVED** [review]
Thank you, @gillisandrew !
LGTM.

Awaiting second LGTM+Approval from any other contributor to this repo before merging.

**Comment by gillisandrew** [comment]
@gmlewis Great! Thanks for your help and feedback, really appreciate it.

**Review by tomfeigin — APPROVED** [review]
Nice!

**Comment by gmlewis** [comment]
Thank you, @tomfeigin !
Merging.

**Pull request merged**
